### PR TITLE
os/blestore/NVMEDevice: fix the I/O logic for read

### PR DESCRIPTION
**Pull request by optimistyzy**
Aio_submit will submit both aio_read/write, and also there
are synchronized read and random read, so we need to
handle the read I/O completion in a correct way.

Also fix the exception logic in command send, make the style
consistent.
 Signed-off-by: optimistyzy <optimistyzy@gmail.com>

### CODE REVIEW

**Comment by optimistyzy** [comment]
This patch is not ready for review, need to be tested. Thanks.

**Comment by liuhongtong** [comment]
I have test this pr based on commit f46b327bb0dd367749397d836c0e090fdfe1a7d4 of master branch.
everything looks well.
seq read: 1 hour
rand read: 18 hours

root@node3:ceph(master)$ ceph -s
2017-03-17 09:52:02.469118 7f98e567b700 -1 WARNING: the following dangerous and experimental features are enabled: bluestore,rocksdb
2017-03-17 09:52:02.471024 7f98e567b700 -1 WARNING: the following dangerous and experimental features are enabled: bluestore,rocksdb
    cluster 5a027377-d1aa-41aa-900c-cb96f2a12e84
     health HEALTH_WARN
            mon.node3 low disk space
     monmap e2: 1 mons at {node3=192.168.200.213:6789/0}
            election epoch 4, quorum 0 node3
        mgr no daemons active 
     osdmap e18: 2 osds: 2 up, 2 in
            flags sortbitwise,require_jewel_osds,require_kraken_osds,require_luminous_osds
      pgmap v24506: 768 pgs, 2 pools, 92232 MB data, 23062 objects
            95042 MB used, 652 GB / 745 GB avail
                 768 active+clean
root@node3:ceph(master)$ ceph df
2017-03-17 09:52:09.907172 7f65cff14700 -1 WARNING: the following dangerous and experimental features are enabled: bluestore,rocksdb
2017-03-17 09:52:09.910367 7f65cff14700 -1 WARNING: the following dangerous and experimental features are enabled: bluestore,rocksdb
GLOBAL:
    SIZE     AVAIL     RAW USED     %RAW USED 
    745G      652G       95042M         12.45 
POOLS:
    NAME     ID     USED       %USED     MAX AVAIL     OBJECTS 
    rbd      0           0         0          608G           0 
    test     1      92232M     12.89          608G       23062 
root@node3:ceph(master)$ ceph osd df
2017-03-17 10:06:45.977757 7fa14b2fc700 -1 WARNING: the following dangerous and experimental features are enabled: bluestore,rocksdb
2017-03-17 10:06:45.983255 7fa14b2fc700 -1 WARNING: the following dangerous and experimental features are enabled: bluestore,rocksdb
ID WEIGHT  REWEIGHT SIZE USE    AVAIL %USE  VAR  PGS 
 0 0.36389  1.00000 372G 44307M  329G 11.61 0.93 373 
 1 0.36389  1.00000 372G 50735M  323G 13.30 1.07 395 
              TOTAL 745G 95042M  652G 12.45          
MIN/MAX VAR: 0.93/1.07  STDDEV: 0.84

**Comment by optimistyzy** [comment]
@liupan1111 @liuhongtong @yuyuyu101  I think that this patch can solve the rand read crash related issue

File: src/os/bluestore/NVMEDevice.cc [hunk 1]
```diff
@@ -1007,14 +1003,11 @@ int NVMEDevice::read_random(uint64_t off, uint64_t len, char *buf, bool buffered
   ++ioc.num_reading;
   driver->queue_task(t);
 
-  {
-    std::unique_lock<std::mutex> l(ioc.lock);
-    while (t->return_code > 0)
-      ioc.cond.wait(l);
+  while (t->return_code > 0) {
+	ioc.aio_wait();
```
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: this isn't a safe check. we may enter the loop while return_code == 0, but before wait other thread change this and signal. at last we are stuck into wait infinitely!

File: src/os/bluestore/NVMEDevice.cc [hunk 2]
```diff
@@ -737,31 +736,31 @@ void io_complete(void *t, const struct spdk_nvme_cpl *completion)
...
-      task->return_code = -1; // FIXME
-    else
-      task->return_code = 0;
     task->fill_cb();
     task->release_segs();
-    {
-      std::unique_lock<std::mutex> l(ctx->lock);
-      ctx->cond.notify_all();
+    // read submitted by AIO
+    if(!task->return_code) {
```
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: why could judge whether is a AIO read by return_code?

Comment (optimistyzy) [inline src/os/bluestore/NVMEDevice.cc]: From the code. For AIO read, we set the initialized value of return_code to  0.

File: src/os/bluestore/NVMEDevice.cc (hunk 1 above)
Comment (optimistyzy) [inline src/os/bluestore/NVMEDevice.cc]: This logic is same with the main  branch code, it only change the ioc.cond.wait(l) with ioc.aio_wait.

**Comment by optimistyzy** [comment]
@yuyuyu101

**Pull request merged**